File: templates/py/UF.py
class UF:
	def __init__(self, n):
		self.parent = [i for i in range(n)]
		self.height = [0 for i in range(n)]
	def find(self, i):
		ancestors = []
		while self.parent[i] != i:
			ancestors.append(i)
			i = self.parent[i]
		for x in ancestors:
			self.parent[x] = i
		return i
	def union(self, i, j):
		x, y = self.find(i), self.find(j)
		if x != y:
			if self.height[x] > self.height[y]:
				self.parent[y] = x
			else:
				self.parent[x] = y
				if self.height[x] == self.height[y]:
					self.height[y] += 1

File: templates/rb/UF.rb
class UF
	attr_accessor :parent, :height
	def initialize(n)
		@parent = Array.new(n) {|i| i}
		@height = Array.new(n, 0)
	end
	def find(i)
		return i if @parent[i] == i
		@parent[i] = find(@parent[i])
		return @parent[i]
	end
	def connected(i, j)
		return find(i) == find(j)
	end
	def union(i, j)
		if !connected(i, j)
			x, y = find(i), find(j)
			if @height[x] > @height[y]
				@parent[y] = x
			else
				@parent[x] = y
				@height[y] += 1 if @height[x] == @height[y]
			end
		end
	end
end
